
From: Marco Martin <notmart@gmail.com>
Date: Fri, 26 May 2017 14:50:10 +0000 (+0200)
Subject: Set the QMenu's window transient parent to the causedPopup's window
X-Git-Url: https://codereview.qt-project.org/gitweb?p=qt%2Fqtbase.git;a=commitdiff_plain;h=0b8f20a8f150eaf2fd688f15333313f4c08d42ea;hp=8e83d6150121412b9f493ff525bce6f09d105463

Set the QMenu's window transient parent to the causedPopup's window

Set the causing widget as the transient parent of the QMenu,
as on Wayland popup surfaces always need a transient parent set
to another surface always currently visible (being their menubar,
parent popup etc

Task-number: QTBUG-60932
Change-Id: I503ed9d28d9a9b4bd70fc3221c0a92cdc923fdc2
---

diff --git a/src/widgets/widgets/qmenu.cpp b/src/widgets/widgets/qmenu.cpp
index 75704f7..91338f1 100644
--- a/src/widgets/widgets/qmenu.cpp
+++ b/src/widgets/widgets/qmenu.cpp
@@ -2303,6 +2303,23 @@ QSize QMenu::sizeHint() const
 void QMenu::popup(const QPoint &p, QAction *atAction)
 {
     Q_D(QMenu);
+    /* Set the causing widget as the transient parent of the QMenu, as on
+       Wayland popup surfaces always need a transient parent set to another
+       surface always currently visible (being their menubar, parent popup etc
+     */
+    if (d->causedPopup.widget && QGuiApplication::platformName().startsWith(QStringLiteral("wayland"))) {
+        // make sure the QWindow for this QMenu widget exists causedpopup's
+        // widget (or its nativeParentwidget if it's a child widget) will
+        // already have a QWindow as it's already visible
+        if (!windowHandle())
+            create(0, true, false);
+
+        QWindow *parentWindow = d->causedPopup.widget->windowHandle();
+        if (!parentWindow)
+            parentWindow = d->causedPopup.widget->nativeParentWidget()->windowHandle();
+        windowHandle()->setTransientParent(parentWindow);
+    }
+
     if (d->scroll) { // reset scroll state from last popup
         if (d->scroll->scrollOffset)
             d->itemsDirty = 1; // sizeHint will be incorrect if there is previous scroll
